chichiwang/exercise-js
# Transpile server with Webpack

## Description
Linked to Issue: #17
Build and watch the server with [Webpack](https://webpack.js.org/) in preparation for setting up a React client to operate with Webpack. This also makes a watch process possible, as well as allowing modular design in the server (TS is unable to handle imports of other TS files).

## Changes
* Update npm scripts:
  * `build:server`: Use webpack to build TS server
  * `start`: Use webpack watcher to reload files on change
* Add devDependencies:
  * `webpack`
  * `webpack-cli`
  * `ts-loader`
  * `webpack-node-externals`
  * `nodemon-webpack-plugin`
* Set up a webpack configuration file `webpack-config.js` to:
  * Build from entry point `server/index.ts`
  * Transpile TypeScript using `ts-loader`
  * Place compiled artifacts in `dist/`
  * Avoid bundling `node_modules/`
  * Avoid bundling Node API
  * Rebuild on change when using the `--watch` flag
    * Ignore test files in the watcher

## Steps to QA
* Pull down and switch to this branch
* Run `npm build:server` and ensure a `dist/index.js` is created
  * Run `node dist/index.js` and ensure the server starts with no issue - verify in browser
* Run `npm start` and ensure Webpack watcher process starts
  * Verify server works in browser
  * Make a change to the response text from the server in `server/index.ts` and ensure the server is rebuilt
  * Verify that the change is reflected in the browser (reload)
  * Make a change to a test file (`server/__tests__/index.test.js` and ensure server is not rebuilt
